File: Discord-Lark/api/gemini_handler.py
# file: api/gemini_handler.py
import os
import google.generativeai as genai

GEMINI_API_KEY = os.getenv('GEMINI_API_KEY')
if not GEMINI_API_KEY:
    raise ValueError("LỖI: GEMINI_API_KEY chưa được thiết lập.")

genai.configure(api_key=GEMINI_API_KEY)

lark_chat_sessions = {}

def get_gemini_response(session_id: str, user_prompt: str) -> str:
    global lark_chat_sessions

    if session_id not in lark_chat_sessions:
        print(f"Bắt đầu phiên trò chuyện mới cho Lark chat ID: {session_id}")
        model = genai.GenerativeModel(
            'gemini-1.5-flash-latest',
            system_instruction="Bạn là một trợ lý AI hữu ích và thân thiện, trả lời các câu hỏi được gửi từ nền tảng Lark. Luôn luôn trả lời bằng tiếng Việt."
        )
        lark_chat_sessions[session_id] = model.start_chat(history=[])

    chat = lark_chat_sessions[session_id]

    try:
        print(f"Đang gửi prompt '{user_prompt}' cho Gemini...")
        response = chat.send_message(user_prompt) # Dùng hàm đồng bộ
        print("Đã nhận được phản hồi từ Gemini.")
        return response.text
    except Exception as e:
        print(f"Lỗi khi gọi Gemini API cho session {session_id}: {e}")
        if session_id in lark_chat_sessions:
            del lark_chat_sessions[session_id]
        return "Xin lỗi, tôi đang gặp sự cố với bộ não Gemini. Vui lòng thử lại sau một lát."

File: api/gemini_handler.js
// file: api/gemini_handler.js
const { GoogleGenerativeAI } = require("@google/generative-ai");

const GEMINI_API_KEY = process.env.GEMINI_API_KEY;
if (!GEMINI_API_KEY) {
    throw new Error("LỖI: GEMINI_API_KEY chưa được thiết lập.");
}

const genAI = new GoogleGenerativeAI(GEMINI_API_KEY);
const larkChatSessions = {};

async function getGeminiResponse(sessionId, userPrompt) {
    if (!larkChatSessions[sessionId]) {
        console.log(`Bắt đầu phiên trò chuyện mới cho Lark chat ID: ${sessionId}`);
        const model = genAI.getGenerativeModel({
            model: 'gemini-1.5-flash-latest',
            systemInstruction: "Bạn là một trợ lý AI hữu ích và thân thiện, trả lời các câu hỏi được gửi từ nền tảng Lark. Luôn luôn trả lời bằng tiếng Việt."
        });
        larkChatSessions[sessionId] = model.startChat({ history: [] });
    }

    const chat = larkChatSessions[sessionId];

    try {
        console.log(`Đang gửi prompt '${userPrompt}' cho Gemini...`);
        const result = await chat.sendMessage(userPrompt);
        const response = result.response;
        console.log("Đã nhận được phản hồi từ Gemini.");
        return response.text();
    } catch (error) {
        console.error(`Lỗi khi gọi Gemini API cho session ${sessionId}:`, error);
        // Xóa phiên bị lỗi để lần sau tạo lại
        delete larkChatSessions[sessionId];
        return "Xin lỗi, tôi đang gặp sự cố với bộ não Gemini. Vui lòng thử lại sau một lát.";
    }
}

module.exports = { getGeminiResponse };